examples/filter.rs

```diff
@@ -0,0 +1,126 @@
+use std::ffi::CString;
+use std::ops::Deref;
+use std::ptr::null_mut;
+use std::sync::RwLock;
+use valkey_module::alloc::ValkeyAlloc;
+use valkey_module::logging::log_notice;
+use valkey_module::{
+    valkey_module, CommandFilter, CommandFilterCtx, Context, RedisModuleCommandFilterCtx,
+    RedisModuleString, RedisModule_CreateString, Status, ValkeyResult, ValkeyString,
+    VALKEYMODULE_CMDFILTER_NOSELF,
+};
+
+// used to register filters using init and deinit, this is more complicated, best use the macro approach below
+static INFO_FILTER: RwLock<Option<CommandFilter>> = RwLock::new(None);
+
+fn init(ctx: &Context, _args: &[ValkeyString]) -> Status {
+    let info_filter = ctx.register_command_filter(info_filter_fn, VALKEYMODULE_CMDFILTER_NOSELF);
+    if info_filter.is_null() {
+        return Status::Err;
+    }
+    let mut info_guard = INFO_FILTER.write().unwrap();
+    *info_guard = Some(info_filter);
+
+    Status::Ok
+}
+
+fn deinit(ctx: &Context) -> Status {
+    let info_guard = INFO_FILTER.read().unwrap();
+    if let Some(ref info_filter) = info_guard.deref() {
+        ctx.unregister_command_filter(info_filter);
+    };
+
+    Status::Ok
+}
+
+// this is used to register and unregister the filter in init and deinit
+// has to be extern "C", better to use the macro approach
+extern "C" fn info_filter_fn(ctx: *mut RedisModuleCommandFilterCtx) {
+    let cf_ctx = CommandFilterCtx::new(ctx);
+
+    // we want the filter to be very efficient as it will be called for every command
+    // check if there is only 1 arg, the command
+    if cf_ctx.args_count() != 1 {
+        return;
+    }
+    // check if cmd (first arg) is info
+    let cmd = cf_ctx.arg_get_as_str(0).unwrap();
+    if !cmd.eq_ignore_ascii_case("info") {
+        return;
+    }
+    // grab client_id
+    let client_id = cf_ctx.get_client_id();
+    log_notice(&format!("info filter for client_id {}", client_id));
+    // replace info with info2 as the command name
+    let custom_cmd = create_module_string("info2");
+    cf_ctx.arg_replace(0, custom_cmd);
+}
+
+// custom command that will replace info command
+fn info2(ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
+    ctx.log_notice("info2 command");
+    //  do something different here
+    Ok("info2".into())
```

Comment:
This currently results in an oddly formatted response:
```
127.0.0.1:7001> info
info2127.0.0.1:7001>
```

Reply:
changed to `info2\n`